File: app/src/main/java/com/dumuzeyn/mp3player/SongRowStateRegistry.java
package com.dumuzeyn.mp3player;

import android.view.View;
import android.os.Trace;
import android.widget.Button;
import android.widget.TextView;
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;

final class SongRowStateRegistry {
    interface StateResolver {
        Track currentTrack();
        boolean isPlaying();
        int activeColor();
        int secondaryActiveColor();
        int inactiveColor();
    }

    private final HashMap<String, Button> playButtons = new HashMap<>();
    private final HashMap<String, View> currentMarkers = new HashMap<>();
    private final HashMap<String, WaveformView> waveforms = new HashMap<>();
    private final HashMap<String, ArrayList<RotatingCoverImageView>> covers = new HashMap<>();
    private final HashMap<String, TextView> titles = new HashMap<>();
    private final HashMap<String, TextView> durations = new HashMap<>();

    void clear() {
        playButtons.clear();
        currentMarkers.clear();
        waveforms.clear();
        covers.clear();
        titles.clear();
        durations.clear();
    }

    void replaceWith(SongRowStateRegistry source) {
        clear();
        playButtons.putAll(source.playButtons);
        currentMarkers.putAll(source.currentMarkers);
        waveforms.putAll(source.waveforms);
        titles.putAll(source.titles);
        durations.putAll(source.durations);
        for (Map.Entry<String, ArrayList<RotatingCoverImageView>> entry : source.covers.entrySet()) {
            covers.put(entry.getKey(), new ArrayList<>(entry.getValue()));
        }
    }

    void forEachCover(CoverConsumer consumer) {
        for (Map.Entry<String, ArrayList<RotatingCoverImageView>> entry : covers.entrySet()) {
            for (RotatingCoverImageView cover : entry.getValue()) {
                consumer.accept(entry.getKey(), cover);
            }
        }
    }

    interface CoverConsumer {
        void accept(String uri, RotatingCoverImageView cover);
    }

    void registerPlayButton(String uri, Button button) {
        playButtons.put(uri, button);
    }

    void registerCurrentMarker(String uri, View marker) {
        currentMarkers.put(uri, marker);
    }

    void registerWaveform(String uri, WaveformView waveform) {
        waveforms.put(uri, waveform);
    }

    void registerCover(String uri, RotatingCoverImageView cover) {
        ArrayList<RotatingCoverImageView> registered = covers.get(uri);
        if (registered == null) {
            registered = new ArrayList<>();
            covers.put(uri, registered);
        }
        if (!registered.contains(cover)) {
            registered.add(cover);
        }
    }

    void registerMetadata(String uri, TextView title, TextView duration) {
        titles.put(uri, title);
        durations.put(uri, duration);
    }

    void refreshMetadata(String uri, Track track, String durationText) {
        TextView title = titles.get(uri);
        if (title != null) {
            title.setText(track.title);
        }
        TextView duration = durations.get(uri);
        if (duration != null) {
            duration.setText(durationText);
        }
    }

    static void applyPlayState(Button button, boolean playing) {
        String symbol = playing ? "\u2161" : "\u25b6";
        if (symbol.contentEquals(button.getText())) return;
        button.setText(symbol);
        int opticalOffset = playing ? 0 : Math.round(
                button.getResources().getDisplayMetrics().density * 2.0f);
        button.setPadding(opticalOffset, 0, 0, 0);
    }

    void refresh(StateResolver resolver) {
        Trace.beginSection("Voltune/Home.refreshSongRows");
        try {
        Track current = resolver.currentTrack();
        String currentUri = current == null ? "" : current.uri;
        boolean playing = resolver.isPlaying();
        for (Map.Entry<String, Button> entry : playButtons.entrySet()) {
            applyPlayState(entry.getValue(), entry.getKey().equals(currentUri) && playing);
        }
        for (Map.Entry<String, View> entry : currentMarkers.entrySet()) {
            int visibility = entry.getKey().equals(currentUri)
                    ? View.VISIBLE : View.INVISIBLE;
            if (entry.getValue().getVisibility() != visibility) {
                entry.getValue().setVisibility(visibility);
            }
        }
        for (Map.Entry<String, WaveformView> entry : waveforms.entrySet()) {
            boolean rowCurrent = entry.getKey().equals(currentUri);
            entry.getValue().setState(
                    rowCurrent ? resolver.activeColor() : resolver.inactiveColor(),
                    resolver.secondaryActiveColor(),
                    rowCurrent && playing);
        }
        for (ArrayList<RotatingCoverImageView> registered : covers.values()) {
            for (RotatingCoverImageView cover : registered) {
                cover.updatePlaybackState(current, playing);
            }
        }
        } finally {
            Trace.endSection();
        }
    }

    void setWaveformsTransitionPaused(boolean paused) {
        for (WaveformView waveform : waveforms.values()) {
            waveform.setTransitionPaused(paused);
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/SongRowStateRegistry.kt
package com.dumuzeyn.mp3player

import android.os.Trace
import android.view.View
import android.widget.Button
import android.widget.TextView
import kotlin.math.roundToInt

internal class SongRowStateRegistry {
    interface StateResolver {
        fun currentTrack(): Track?
        fun isPlaying(): Boolean
        fun activeColor(): Int
        fun secondaryActiveColor(): Int
        fun inactiveColor(): Int
    }

    fun interface CoverConsumer {
        fun accept(uri: String, cover: RotatingCoverImageView)
    }

    private val playButtons = HashMap<String, Button>()
    private val currentMarkers = HashMap<String, View>()
    private val waveforms = HashMap<String, WaveformView>()
    private val covers = HashMap<String, ArrayList<RotatingCoverImageView>>()
    private val titles = HashMap<String, TextView>()
    private val durations = HashMap<String, TextView>()

    fun clear() {
        playButtons.clear()
        currentMarkers.clear()
        waveforms.clear()
        covers.clear()
        titles.clear()
        durations.clear()
    }

    fun replaceWith(source: SongRowStateRegistry) {
        clear()
        playButtons.putAll(source.playButtons)
        currentMarkers.putAll(source.currentMarkers)
        waveforms.putAll(source.waveforms)
        titles.putAll(source.titles)
        durations.putAll(source.durations)
        source.covers.forEach { (uri, registered) -> covers[uri] = ArrayList(registered) }
    }

    fun forEachCover(consumer: CoverConsumer) {
        covers.forEach { (uri, registered) ->
            registered.forEach { cover -> consumer.accept(uri, cover) }
        }
    }

    fun registerPlayButton(uri: String, button: Button) {
        playButtons[uri] = button
    }

    fun registerCurrentMarker(uri: String, marker: View) {
        currentMarkers[uri] = marker
    }

    fun registerWaveform(uri: String, waveform: WaveformView) {
        waveforms[uri] = waveform
    }

    fun registerCover(uri: String, cover: RotatingCoverImageView) {
        val registered = covers.getOrPut(uri, ::ArrayList)
        if (!registered.contains(cover)) registered.add(cover)
    }

    fun registerMetadata(uri: String, title: TextView, duration: TextView) {
        titles[uri] = title
        durations[uri] = duration
    }

    fun refreshMetadata(uri: String, track: Track, durationText: String) {
        titles[uri]?.text = track.title
        durations[uri]?.text = durationText
    }

    fun refresh(resolver: StateResolver) {
        Trace.beginSection("Voltune/Home.refreshSongRows")
        try {
            val current = resolver.currentTrack()
            val currentUri = current?.uri.orEmpty()
            val playing = resolver.isPlaying()
            playButtons.forEach { (uri, button) ->
                applyPlayState(button, uri == currentUri && playing)
            }
            currentMarkers.forEach { (uri, marker) ->
                val visibility = if (uri == currentUri) View.VISIBLE else View.INVISIBLE
                if (marker.visibility != visibility) marker.visibility = visibility
            }
            waveforms.forEach { (uri, waveform) ->
                val rowCurrent = uri == currentUri
                waveform.setState(
                    if (rowCurrent) resolver.activeColor() else resolver.inactiveColor(),
                    resolver.secondaryActiveColor(),
                    rowCurrent && playing,
                )
            }
            covers.values.forEach { registered ->
                registered.forEach { cover ->
                    cover.updatePlaybackState(current, playing)
                }
            }
        } finally {
            Trace.endSection()
        }
    }

    fun setWaveformsTransitionPaused(paused: Boolean) {
        waveforms.values.forEach { it.setTransitionPaused(paused) }
    }

    companion object {
        @JvmStatic
        fun applyPlayState(button: Button, playing: Boolean) {
            val symbol = if (playing) "Ⅱ" else "▶"
            if (symbol.contentEquals(button.text)) return
            button.text = symbol
            val opticalOffset = if (playing) {
                0
            } else {
                (button.resources.displayMetrics.density * 2f).roundToInt()
            }
            button.setPadding(opticalOffset, 0, 0, 0)
        }
    }
}
